### ISSUE TITLE
Is it already possible to load text file contents into JavaScript?

### ISSUE THREAD

**Issue by jacksonrayhamilton** [rails/sprockets#471]
I have an SVG file that I want to dynamically load, inline, via JavaScript in my Rails application.  I need to load the SVG inline so that I can apply a CSS `fill` property to it.

I'd rather not load the SVG file via AJAX to avoid potential lag before the SVG is visible.

Ideally what I'd like to do is to tell Sprockets that, in my `app/assets/icons/` directory, I have icons `calendar.svg` and `phone.svg`, and that Sprockets should take the contents of those SVG files and set them as string values in a JavaScript object `SVG`, so that I may access them in my JavaScript code like this:

    var view = {
      render: function () {
        this.element.innerHTML = SVG['app/assets/icons/calendar.svg'];
      }
    };

Preferably, the solution is generalized to more than just SVGs, and I could potentially load any data in text format via JS, e.g. `TEXT['app/assets/icons/calendar.svg']` or even `TEXT['app/assets/etc/some-text-file.text']`.

Is there already some way to do this in Sprockets, or do I need to write a plugin for it?

**Comment by vincentwoo** [comment]
Yes, you can use ERB and the data url helpers. Something like:

```js
$('<img>').attr('src', '<%= asset_data_uri 'avatar.jpg' %>')
```

should work. See https://github.com/rails/sprockets/blob/a9c65250f46429912dbbab2859dcb08bb5c32f95/lib/sprockets/context.rb#L197

**Comment by jacksonrayhamilton** [comment]
By suggesting ERB, I believe that you are suggesting I echo that data URI in the HTML that is initially returned from my server.  However, this won't work particularly well for my application.  

My Rails application has a single, simple view. An `application.js` file is generated via the asset pipeline.  This `application.js` makes the application a SPA, because it dynamically generates the HTML for the page.

If I have lots of SVGs in the HTML returned from my server, it will take longer for the browser to finish downloading the HTML and display the page; whereas if I load the SVGs in one network request (perhaps the JS request that I already have to make for `application.js`), then that loading can be deferred until after there is something on the page for the user to look at.  So I think the solution for me will be to include the text file's contents in my `application.js`.  Assuming that is how I should solve this, can I do that?

**Comment by vincentwoo** [comment]
Your question is very poorly explained, I think. What is this business with "text files"? You can either inline your image data in your JS payload or not. It's up to you whether you want to make that tradeoff. 

I read your repo description and it was nonsensical to me: 

> Why do this? Your Rails application might serve up a single-page web application, which requires text assets (like SVGs). You need these assets to render your application, and you don't want to fetch them via AJAX (to avoid lag), nor embed them in the returned HTML (to keep the HTML response small).

You have to download your asset at *some point*, probably one of:

1. up front, in the application bundle
2. immediately after app load
3. as the asset is required

Whether you do any of these has very little to do with whether the asset in question is "text" or an image.

**Comment by jacksonrayhamilton** [comment]
An SVG has a textual representation, e.g.,

```svg
<svg viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg">
  <circle cx="100" cy="100" r="100"/>
</svg>
```

Therefore, it is one example of a "text file."  It happened to be the only type that I cared about here.

However, there's no telling whether I might want to load some ".md", ".txt", or other types of "text files" files down the line, so I used open-ended language to imply that the solution should be extensible to this type of file.  I limited the definition to "text files", since while I could load "binary files" like ".jpeg", SQL databases, etc as well, those are less practical to translate to JavaScript strings, since they'll probably be base64-encoded and become much larger in size.
